webui/react/src/modal/useModality.tsx

```diff
@@ -0,0 +1,183 @@
+import { ExclamationCircleOutlined } from '@ant-design/icons';
+import { Modal as AntdModal } from 'antd';
+import React, {
+  createContext,
+  Dispatch,
+  ReactNode,
+  SetStateAction,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
+
+import Button from 'components/kit/Button';
+import Link from 'components/Link';
+import Icon from 'shared/components/Icon';
+
+import css from './useModality.module.scss';
+
+interface LinkParams {
+  text: string;
+  url: string;
+}
+
+type ModalSize = 'small' | 'medium' | 'large';
+
+export interface ModalSubmitParams {
+  disabled?: boolean;
+  text: string;
+  handler: () => Promise<void>;
+  onComplete?: () => Promise<void>;
+}
+
+export interface UseModalParams {
+  cancelText: string;
+  danger?: boolean;
+  footerLink?: LinkParams;
+  headerLink?: LinkParams;
+  icon?: string;
+  key?: string;
+  size: ModalSize;
+  submit: ModalSubmitParams;
+  titleText: string;
+}
+
+interface ModalContext {
+  params?: UseModalParams;
+  setParams: Dispatch<SetStateAction<UseModalParams | null>>;
+}
+
+const ModalContext = createContext<ModalContext | null>(null);
+
+export function useModalParams(params: UseModalParams): void {
+  const modalContext = useContext(ModalContext);
+  if (modalContext === null) throw new Error('tried to use modal context outside of modal');
+  const { setParams } = modalContext;
+  useEffect(() => setParams(params), [setParams, params]);
+}
+
+interface ModalContainerProps {
+  modalIsOpen: boolean;
+  setModalIsOpen: Dispatch<SetStateAction<boolean>>;
+  children: ReactNode;
+}
+
+function ModalContainer({ modalIsOpen, children: modalBody, setModalIsOpen }: ModalContainerProps) {
+  /*
+   * the params are defined here, but set from with the ModalBodyComponent
+   * via useModalParams (which uses React Context)
+   */
+  const [params, setParams] = useState<UseModalParams | null>(null);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const close = useCallback(() => setModalIsOpen(false), [setModalIsOpen]);
+
+  const handleOk = useCallback(async () => {
+    setIsSubmitting(true);
+    await params?.submit?.handler();
```

Comment:
The current pattern in the case of a submission failure is to close the modal and show a toast notification (the modal component itself should have already done all the validation it can).

The logic to show the toasts is currently in the various handle submit functions. As it, this would maintain that.

Reply:
I recognize this is the current design pattern but it's not really a good one. I'd like to open the possibility of handling the error without closing the modal.

Reply:
in that have do you have a preference as to the API? 

one option would be to have the `handleSubmit` not catch their errors and then have a `handleSubmitError (e: Error) => void`, although this would add another quirk to the way the creators of specific modal components have to do things.

Reply:
Generally my preference is to expose callbacks but if we wanted to keep this promise-based approach we could just handle a failed promise.